zebra-chain/src/transaction/sighash.rs

```diff
@@ -0,0 +1,924 @@
+use super::Transaction;
+use crate::{
+    parameters::{ConsensusBranchId, NetworkUpgrade},
+    serialization::{WriteZcashExt, ZcashSerialize},
+    transparent,
+};
+use blake2b_simd::Hash;
+use byteorder::{LittleEndian, WriteBytesExt};
+use io::Write;
+use std::io;
+
+// TODO: figure out how to write this more concisely
+static ZIP243_EXPLANATION: &str =
+    "The new algorithm MUST be used for signatures created over the Sapling
+transaction format 2. Combined with the new consensus rule that v3
+transaction formats will be invalid from the Sapling upgrade, this
+effectively means that all transaction signatures from the Sapling activation
+height (as specified in 6) will use the new algorithm.";
+
+const OVERWINTER_VERSION_GROUP_ID: u32 = 0x03C4_8270;
+const SAPLING_VERSION_GROUP_ID: u32 = 0x892F_2085;
+
+const ZCASH_SIGHASH_PERSONALIZATION_PREFIX: &[u8; 12] = b"ZcashSigHash";
+const ZCASH_PREVOUTS_HASH_PERSONALIZATION: &[u8; 16] = b"ZcashPrevoutHash";
+const ZCASH_SEQUENCE_HASH_PERSONALIZATION: &[u8; 16] = b"ZcashSequencHash";
+const ZCASH_OUTPUTS_HASH_PERSONALIZATION: &[u8; 16] = b"ZcashOutputsHash";
+const ZCASH_JOINSPLITS_HASH_PERSONALIZATION: &[u8; 16] = b"ZcashJSplitsHash";
+const ZCASH_SHIELDED_SPENDS_HASH_PERSONALIZATION: &[u8; 16] = b"ZcashSSpendsHash";
+const ZCASH_SHIELDED_OUTPUTS_HASH_PERSONALIZATION: &[u8; 16] = b"ZcashSOutputHash";
+
+bitflags::bitflags! {
+    pub struct HashType: u32 {
+        const ALL = 0b0000_0001;
+        const NONE = 0b0000_0010;
+        const SINGLE = Self::ALL.bits | Self::NONE.bits;
+        const ANYONECANPAY = 0b1000_0000;
+    }
+}
+
+impl HashType {
+    fn masked(self) -> Self {
+        Self::from_bits_truncate(self.bits & 0b0001_1111)
+    }
+}
+
+pub(super) struct SigHasher<'a> {
+    trans: &'a Transaction,
+    hash_type: HashType,
+    network_upgrade: NetworkUpgrade,
+    input: Option<(transparent::Output, &'a transparent::Input, usize)>,
+}
+
+impl<'a> SigHasher<'a> {
+    pub fn new(
+        trans: &'a Transaction,
+        hash_type: HashType,
+        network_upgrade: NetworkUpgrade,
+        input: Option<(u32, transparent::Output)>,
+    ) -> Self {
+        let input = if let Some((index, prevout)) = input {
+            let index = index as usize;
+            let inputs = trans.inputs();
+
+            Some((prevout, &inputs[index], index))
+        } else {
+            None
+        };
+
+        SigHasher {
+            trans,
+            hash_type,
+            network_upgrade,
+            input,
+        }
+    }
+
+    pub fn sighash(self) -> Hash {
```

Comment:
Panics if the network upgrade is Genesis or BeforeOverwinter, or the transaction is v1 or v2.

Reply:
I'm gonna put this documentation on the `sighash` fn on `Transaction` and instead restrict the visibility on this to `pub(super)`